---
title: Bioinformatic Pipeline - "Hiding in Plain Sight - Genome-wide recombination
  and horizontal gene transfer drive clonal diversity in Fusarium oxysporum fsp ciceris"
author: "Fayyaz A, Robinson G, Chang PL, Bekele D, Yimer SM, Carrasquilla-Garcia N, Negash K, Anand K. Surendrarao, Eric J.B. von Wettberg, Robbertse B, Seid Ahmedi, Tesfaye K, Fikre Ab, Farmer AD and Cook DR"
#date: '2023-02-07'
output:
  #rmdformats::readthedown:
  #  toc_depth: 5
  #  self_contained: true
  #  thumbnails: false
  #  lightbox: true
  #  gallery: true
  #  highlight: tango
  html_document: default
  #pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#library(reticulate)
```

---

## Contents

### 1. [Overview](#Overview)
### 2. [Genome Assembly](#assembly)
### 3. [Gene Annotation and Analyses](#gene_annotation)
### 4. [Identification of SNP and Genetic Groups](#read_mapping)
### 5. [Population Analyses](#pop_gen)
### 6. [References](#refs)

---

## Help with Rmarkdown
### https://holtzy.github.io/Pimp-my-rmd/
### https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf

## <a id="Overview"></a>1. Overview

Plot overview pipeline

## <a id="assembly"></a>2. Genome Assembly

```{r, eval=F}
#####################################################################################
######################### Need input from Amna/Peter ################################
#####################################################################################
```

A. Quality check with FastQC
B. Low quality reads and adapters trimmed using Trimmomatic v36
C. Error correction using ALLPATHS-LG
D. Assembly using SPAdes
E. Genome completeness assessed using BUSCO v3 (Ascomycota odb9 dataset)

## <a id="gene_annotation"></a>3. Gene Annotation and Analyses

```{r, eval=F}
#####################################################################################
######################### Need input from Amna/Peter ################################
#####################################################################################
```

A. Average nucleotide identity calculated with Pyani
```{bash, eval=False}
/home/localhost/bin/singularity run /home/localhost/singularityImages/average_nucleotide_identity-v0.2.9.simg -i all-file-pyani/ -g -m ANIb --workers 20 -o pyani-results
```
B. Annotation
  - RepeatModeler
  - MAKER
  - Interpro, GeneOntology, AHRD
  - Gene Enrichment with BLAST2GO
  - SIX genes homologs BLASTp
  - Orthologs identified with Orthofinder (default excpt with -msa)
C. Phylogenetic Analyses
  - BEAST
  - iTOL
  - Robinson-Foults with ETE3 toolkit
D. Pangenome
  - PanOCT

## <a id="read_mapping"></a>4. Identification of SNP and genetic groups

### A. Read mapping

### i. Illumina read mapping was conducted with BWA MEM 0.7.9a-r786

Example -- https://www.illumina.com/products/by-type/informatics-products/basespace-sequence-hub/apps/bwa-aligner.html

Tools required:

- BWA MEM v0.7.9a-r786

- SAMtools v1.3.1

- Picard 

```{bash, eval=FALSE}
#create index for F. oxysporum reference genome (Fol4287 - GCF_000149955.1)
bwa index [-a bwtsw|is] reference.fasta index_prefix

#running BWA (looped through all Illumina paired-end read data)
#loops through Illumina read prefixes 
for f in $(ls SRR* | cut -d"_" -f1); \
do bwa mem -M -t 32 reference.fasta -P ${f}_1.fastq ${f}_2.fastq > ${f}.sam; \
done

#converts SAM file to BAM file
for f in $(ls *.sam | cut -d"." -f1); do samtools fixmate -O bam ${f}.sam ${f}.bam
```

### ii. Variant calling with GATK v4.1 against Fol4287 reference

Tools required:

- GATK v4.1

```{bash, eval=F}
#####################################################################################
######################### Need input from Amna/Peter ################################
#####################################################################################
```


### B. Filtering VCF file 

Tools required:

bcftools/1.16

vcftools/b240116

For most population analyses, we exclusively used high quality genomes (based on BUSCO completeness > 97%, N50, number of scaffolds and genome size) - see "Genome Assembly". This provided a total list of 120 genomes, 99 of which were taxonomically within the FOSC (Fig. 2). This is reduced from the earlier VCF file, which contained 320 isolates (Supplemental Fig. 1)

Reducing the VCF file from 320 to 120, was done using bcftools (custom bash script).

First, we split the VCF into separate single isolate VCF files

```{bash, eval=F}
######## ./splitting_vcfs_retaining.sh ###########

#!/usr/bin/env bash
for file in *.vcf.gz; do
  for sample in `bcftools query -l $file`; do
    bcftools view -Oz -s $sample -o ${file/.vcf*/.$sample.vcf.gz} $file
  done
done
```

Second, we indexed all the single isolate VCF files

```{bash, eval=F}
for f in *vcf.gz; do tabix -f -p vcf $f; done
```

Third, we merged isolates from list of high quality genomes (information obtained from assembly data)

```{bash, eval=F}
bcftools merge --file-list isolate_list > ethiopia_fo_structure.vcf
```

To ensure that the subsequent VCF was high quality, especially for linkage disequilibrium, we adjusted filters to learn how different filters affect the subsequent linkage disequilibrium analyses. 

The main two parameters were missingness and minor allele frequency (MAF). MAF was of particular importance to us, as often researchers adjust MAF cutoff as >0.05 without regard for the effects on the population. 

```{bash, eval=F}
#VCF 
vcftools --vcf ethiopia_fo_structure.vcf --maf 0.05 --max-missing 0.9 --recode --stdout | bgzip -c > all_ethiopia_isolates.0_05.vcf.gz

#note that this was also conducted but there were only minor differences in the outcomes. The overall clonal groupings remained the same. The LD decay curve showed minor differences

#include plot /home/guy/Documents/Research/Cook/fusarium_pnas_final/old/Linkage/ld_total_maf.png
```


To do this we 
  - missingness
  - maf
  - heterozygous
  - masking ambiguous SNP calls
  - Mapping informed with MUMmer whole genome alignments
  
C. Identification of clonal groups
  - Subsampling SNPs ~10% X10 times
  - STRUCTURE
  - STRUCTUREharvester - Evanno stats
  - CLUMPP v1.1.2 merge all 10 replicates (LargeKgreedy)
  - STRUCTURE plotted with R script
  - Assigning clonal groupings using Euclidean distances

## <a id="pop_gen"></a>5. Population Analyses

A. Geographic Analyses
  - Plotting onto maps in R
  - Mantel tests
B. Linkage disequilibrium
  - Additional filtering for derived SNPs and ancestral SNPs
    - BASH scripts
  - LD with r2 (PLINK)
    - Inter and intra-chromosomal LD
    - Genome-wide LD across genome
  - LD with rbarD (Poppr)
C. Nucleotide Diversity (pi)
D. Tajima's D
E. Fst (Weighted)
  - Core SNP Fst
  - Pangenome Fst
  - BUSCO Fst

## <a id="refs"></a>6. References